lib/rust/frp/src/stream.rs

```diff
@@ -14,45 +14,53 @@ use crate::data::watch;
 
 /// Owned call stack type.
 #[cfg(feature = "stack-trace")]
-pub type OwnedCallStack = EnabledCallStack;
+pub type OwnedCallStack<'a> = EnabledCallStack<'a>;
 
 /// Owned call stack type.
 #[cfg(not(feature = "stack-trace"))]
-pub type OwnedCallStack = DisabledCallStack;
+pub type OwnedCallStack<'a> = DisabledCallStack;
 
 /// A call stack trace for FRP events.
-pub type CallStack<'a> = &'a OwnedCallStack;
+pub type CallStack<'a> = &'a OwnedCallStack<'a>;
 
 
 // === Ops ===
 
 /// Call stack operations available on both enabled and disabled stack implementations.
-pub trait CallStackOps: Default + Display {
+pub trait CallStackOps<'a>: Default + Display {
     /// Create a sub stack trace.
-    fn sub(&self, label: Label) -> Self;
+    fn sub(&'a self, label: Label) -> Self;
```

Comment:
Can you explain to me why need to specify `'a` here? Is there some implicit that causes returned Self to be also constrained by `'a`?

Reply:
It is necessary to tie the input self ref lifetime with the trait, so then the `Self` type in any implementation can store references to that lifetime and let them return from `sub`. Without that tie, you would get a lifetime error in trait implementation.
```
43 |     fn sub(&self, label: Label) -> Self {
   |            - let's call the lifetime of this reference `'1`
44 |         Self { parent: Some(self), label }
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ associated function was supposed to return data with lifetime `'a` but it is returning data with lifetime `'1`
   ```

Reply:
Ok, so trait parameter of the trait means that the structure implementing it is constraint by it. It wasn't obvious for me.

Reply:
It doesn't have to be necessarily, but it can. You are free to use that `'a` in the implementation however you want (e.g. in impl bounds), as long as the trait interface is upheld in the end. In this case, the `DisabledCallStack` struct doesn't use it beyond the trait required interface, but `EnabledCallStack` binds its own lifetime arg to the trait, so that the desired `sub` implementation is possible.